rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,716 @@
+lengthened. We propose raising this limit to 7 days, but keeping the default
+(1hr) the same.
+
+#### Bot Instance Locking
+
+When we introduced bot instances in
+[RFD0162](0162-machine-id-token-join-method-bot-instance.md), it allowed many
+bot instances to join under a single bot (and Teleport user). Generation counter
+checks were moved out of a bot's user and into bot instances, however when a
+generation counter mismatch occurs, the resulting lock is still filed against
+the user as a whole. This means a generation counter lockout in one instance can
+easily impact all other instances under the same bot.
+
+We can introduce several new lock targets to address this:
+
+- Bot instance UUID locking: prevent access by a particular instance. This will
+  not lock bots that have since reauthenticated and received a new bot instance
+  UUID.
+
+- Join token locking: locks all bot instances that joined with a particular
+  token. This may require introduction of a new certificate field to track the
+  exact join token used.
+
+- Public key locking: locks bots joining with a particular public key. A
+  compromised bot could theoretically generate a fresh keypair so a join token
+  lock is the primary locking solution, however this will prevent joining with
+  another token.
+
+#### Bot Joining URIs
+
+Instead of separate flags for proxy, join method, token value, and other
+potential join-method specific parameters, we propose adopting joining URIs.
+These would provide a single value users can copy to pre-fill various
+configuration fields and would greatly improve the onboarding experience.
+
+The URI syntax might look like this:
+```
+tbot+[auth|proxy]://[join method]:[token value]@[addr]:[port]?key=val&foo=bar
+```
+
+Consider these two equivalent commands:
+```
+$ tbot start identity --proxy-server example.teleport.sh:443 --join-method bound-keypair --token example
+
+$ tbot start identity tbot+proxy://bound-keypair:example@example.teleport.sh:443
+```
+
+Joining URIs can greatly simplify the regular onboarding experience by providing
+a single value to copy when onboarding a bot:
+
+```
+$ tctl bots add example --roles=access
+The bot token: tbot+proxy://bound-keypair:example@example.teleport.sh:443
+This token will expire in 59 minutes.
+
+[...snip...]
+
+$ tbot start identity tbot+proxy://bound-keypair:example@example.teleport.sh:443 ...
+```
+
+Given the CLI now supports many operational modes, it's much easier for users to
+write their given starting command (e.g. `tbot start app`) and paste the joining
+URI to get started immediately.
+
+URL paths and query parameters may also provide options for future extension if
+desired.
+
+## Future Extensions and Alternatives
+
+### Agent Joining Support
+
+We should explore expanding this join method to cover regular Teleport agent
+joining as well as bots, as a more secure alternative to static or long-lived
+join tokens.
+
+### Additional Keypair Protections
```

Comment:
Would you allow CA based joining? In that case someone could pregenerate the tbot private keypair and sign it from a centralized CA.
One could then add that public CA to the ProvisionToken as allowed to rejoin.

That way all tbot's could rejoin if they have a certificate signed by a single matching CA.